### PR TITLE
cam6_4_50:  Implement CCPPized check_energy_chng and check_energy_fix

### LINKED ISSUE #1138: Conversion of check_energy to CCPP
### Issue Type

Infrastructure Update

### Issue Description

Move `src/physics/cam/check_energy.F90` to atmospheric_physics and CCPP-ize:
* `check_energy_chng`
* `check_energy_fix`

Relevant support subroutines, e.g., `src/utils/gmean_mod.F90` will be de-chunkized and copied to CAM-SIMA as necessary.

Companion PR: https://github.com/ESCOMP/atmospheric_physics/issues/114

### Will this change answers?

No

### Will you be implementing this yourself?

Yes

### PR DESCRIPTION
**Pull request by jimmielin**
Closes #1138

## Purpose of changes
- CCPPize check_energy_chng and check_energy_fix
  CAM interfaces have been kept in check_energy.F90 instead of a new module because check_tracers is not yet CCPPized, there are other non-CCPPized routines in module, and for compatibility with FV3 external calls
- Save air_composition cp_or_cv_dycore into state snapshot
- Separate out "energy_formula_physics"/"energy_formula_dycore" definitions used in get_hydrostatic_energy from dyn_tests_utils "vcoord" for an eventual change in SIMA.

## Related issues
- Companion PR in atmos_phys: https://github.com/ESCOMP/atmospheric_physics/pull/145
- Companion PR in CAM-SIMA: https://github.com/ESCOMP/CAM-SIMA/pull/316

Files changed:
```
List all files eliminated: N/A

List all files added and what they do:
- separate out energy formula definitions used in get_hydrostatic_energy from dyn_tests_utils "vcoord"
A       src/utils/cam_thermo_formula.F90

List all existing files that have been modified, and describe the changes:

Update atmos_phys external with CCPPized check_energy
M       .gitmodules
M       bld/configure
M       src/atmos_phys

Save cp_or_cv_dycore in state snapshot
M       src/control/cam_snapshot_common.F90

New CAM interfaces to CCPPized routines, when avaialble
M       src/physics/cam/check_energy.F90

Update calls to CCPPized routines, when available
M       src/physics/cam/physpkg.F90
M       src/physics/cam/zm_conv_intr.F90
M       src/physics/cam7/physpkg.F90
M       src/physics/simple/physpkg.F90
M       src/physics/spcam/crm_physics.F90
M       src/physics/spcam/spcam_drivers.F90
```

### CODE REVIEW

**Review by cacraigucar — CHANGES_REQUESTED** [review]

File: .gitmodules [hunk 1]
```diff
@@ -35,8 +35,8 @@
 
 [submodule "atmos_phys"]
 	path = src/atmos_phys
-	url = https://github.com/ESCOMP/atmospheric_physics
-        fxtag = atmos_phys0_06_000
+	url = https://github.com/jimmielin/atmospheric_physics
+        fxtag = b1e2c2a9
```
[the remark this answers is not in the document]
Comment (jimmielin) [reply .gitmodules]: Thanks, updated to `atmos_phys0_07_000`!

File: src/physics/cam/check_energy.F90 [hunk 2]
```diff
@@ -982,5 +558,358 @@ subroutine tot_energy_phys(state, outfld_name_suffix,vc)
...
+    use air_composition,    only: cpairv, cp_or_cv_dycore
+
+    ! CCPP-ized subroutine
+    use check_energy_chng,  only: check_energy_chng_run
+
+    type(physics_state), intent(inout) :: state
+    type(physics_tend ), intent(inout) :: tend
+    character*(*),intent(in) :: name               ! parameterization name for fluxes
+    integer , intent(in) :: nstep                  ! current timestep number
+    real(r8), intent(in) :: ztodt                  ! 2 delta t (model time increment)
```
Comment (nusbaume) [inline src/physics/cam/check_energy.F90:761]: I would change this variable comment to the following:

```suggestion
    real(r8), intent(in) :: ztodt                  ! physics timestep (s)
```

File: src/physics/cam/check_energy.F90 [hunk 3]
```diff
@@ -982,5 +558,358 @@ subroutine tot_energy_phys(state, outfld_name_suffix,vc)
...
+    real(r8), intent(in) :: flx_vap(:)             ! (pcols) - boundary flux of vapor         (kg/m2/s)
+    real(r8), intent(in) :: flx_cnd(:)             ! (pcols) -boundary flux of liquid+ice    (m/s) (precip?)
+    real(r8), intent(in) :: flx_ice(:)             ! (pcols) -boundary flux of ice           (m/s) (snow?)
```
Comment (nusbaume) [inline src/physics/cam/check_energy.F90:764]: I might change these variable comments to the following:

```suggestion
    real(r8), intent(in) :: flx_cnd(:)             ! (pcols) -boundary flux of lwe liquid+ice  (m/s)
    real(r8), intent(in) :: flx_ice(:)             ! (pcols) -boundary flux of lwe ice         (m/s) 
```

File: src/physics/cam/check_energy.F90 [hunk 4]
```diff
@@ -982,5 +558,358 @@ subroutine tot_energy_phys(state, outfld_name_suffix,vc)
...
+        ! No subcolumns
+        local_cp_phys(:ncol,:) = cpairv(:ncol,:,lchnk)
+
+        ! Only if using MPAS or SE energy formula cp_or_cv_dycore is nonzero.
+        if(vc_dycore == ENERGY_FORMULA_DYCORE_MPAS .or. vc_dycore == ENERGY_FORMULA_DYCORE_SE) then
+            local_cp_or_cv_dycore(:ncol,:) = cp_or_cv_dycore(:ncol,:,lchnk)
+
+            scaling_dycore(:ncol,:)  = cpairv(:ncol,:,lchnk)/local_cp_or_cv_dycore(:ncol,:) ! cp/cv scaling
+        endif
+    elseif(state%psetcols > pcols) then
```
Comment (nusbaume) [inline src/physics/cam/check_energy.F90:788]: Total nit-pick, but I would add a space here:

```suggestion
    else if(state%psetcols > pcols) then
```

**Review by nusbaume — CHANGES_REQUESTED** [review]
Looks good!  Just had a few requests, almost entirely related to comments.

File: src/physics/simple/physpkg.F90 [hunk 5]
```diff
@@ -831,9 +832,9 @@ subroutine tphysbc (ztodt, state, tend, pbuf, cam_out, cam_in )
     call t_startf('energy_fixer')
 
     if (adiabatic .and. (.not. dycore_is('EUL'))) then
-      call check_energy_fix(state, ptend, nstep, flx_heat)
+      call check_energy_cam_fix(state, ptend, nstep, flx_heat)
       call physics_update(state, ptend, ztodt, tend)
-      call check_energy_chng(state, tend, "chkengyfix", nstep, ztodt, zero, zero, zero, flx_heat)
+      call check_energy_cam_chng(state, tend, "chkengyfix", nstep, ztodt, zero, zero, zero, flx_heat)
       call outfld( 'EFIX', flx_heat    , pcols, lchnk   )
```
[the remark this answers is not in the document]
Comment (jimmielin) [reply src/physics/simple/physpkg.F90:838]: Thanks, made an issue https://github.com/ESCOMP/atmospheric_physics/issues/161

One needs to be renamed and some need to be added/computed in a separate diagnostic scheme.

File: src/control/cam_snapshot_common.F90 [hunk 6]
```diff
@@ -843,6 +847,9 @@ subroutine state_snapshot_all_outfld(lchnk, file_num, state)
       case ('state%te_cur_dyn')
          call outfld(state_snapshot(i)%standard_name, state%te_cur(:, dyn_te_idx), pcols, lchnk)
 
+      case ('air_composition_cp_or_cv_dycore')
+         call outfld(state_snapshot(i)%standard_name, cp_or_cv_dycore(:,:,lchnk), pcols, lchnk)
+
```
[the remark this answers is not in the document]
Comment (jimmielin) [reply src/control/cam_snapshot_common.F90:852]: Thanks, added!

File: src/physics/cam/check_energy.F90 (hunk 2 above)
Comment (jimmielin) [reply src/physics/cam/check_energy.F90:761]: Thanks, updated!

File: src/physics/cam/check_energy.F90 (hunk 3 above)
Comment (jimmielin) [reply src/physics/cam/check_energy.F90:764]: Thanks, updated!

File: src/physics/cam/check_energy.F90 (hunk 4 above)
Comment (jimmielin) [reply src/physics/cam/check_energy.F90:788]: Thanks, updated!

File: src/physics/cam/check_energy.F90 [hunk 7]
```diff
@@ -892,7 +468,7 @@ subroutine tot_energy_phys(state, outfld_name_suffix,vc)
     end if
 
     if (state%psetcols == pcols) then
-      if (vc_loc == vc_height .or. vc_loc == vc_dry_pressure) then
+      if (vc_loc == ENERGY_FORMULA_DYCORE_MPAS .or. vc_loc == ENERGY_FORMULA_DYCORE_SE) then
```
[the remark this answers is not in the document]
Comment (jimmielin) [reply src/physics/cam/check_energy.F90:474]: Thanks, added above where `vc_loc` is first set!

**Review by nusbaume — APPROVED** [review]
Looks great to me now.  Thanks @jimmielin!

**Review by cacraigucar — APPROVED** [review]

**Comment by jimmielin** [comment]
Thanks @nusbaume and @cacraigucar for the code reviews! I will wait for my turn to make a tag and run the test suite again.

If before my turn to make a tag, https://github.com/ESCOMP/atmospheric_physics/pull/160 is incorporated in a atmos_phys tag, I will use that tag and update the subroutine call for `check_energy_fix_run` to include dummy arguments for `errmsg` and `errflg`. Otherwise I will keep this on `atmos_phys0_07_000`.

**Pull request merged**